gluten-core/src/main/java/org/apache/gluten/memory/memtarget/spark/DynamicOffHeapSizingPolicyChecker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.memory.memtarget.spark;
+
+import org.apache.gluten.GlutenConfig;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public final class DynamicOffHeapSizingPolicyChecker {
```

Comment:
Will investigate.

Reply:
Can you please elaborate your suggestion a bit? The approach here leans on the policy checker being static, accounting for used off-heap bytes across all memory targets (say, if memory isolation is on). Are you suggesting a new TreeMemoryTarget or something that is an alternative to it, which presumably would have similar complexity since we have to support, say, memory isolation?

Reply:
> Are you suggesting a new TreeMemoryTarget or something that is an alternative to it, which presumably would have similar complexity since we have to support, say, memory isolation?

We don't have to add something as heavy as TreeMemoryTarget. If I understand correctly, `DynamicOffHeapSizingPolicyChecker` was designed to be a global singleton, right? If that's true, we can make the policy checker extend API https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/java/org/apache/gluten/memory/memtarget/MemoryTarget.java#L23 and having the allocations accounted into a global static counter in policy checker.

Just like the logging memory target https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/java/org/apache/gluten/memory/memtarget/LoggingMemoryTarget.java, it logs allocation through the static `LOGGER` object, and delegates all the allocate to the subsequent target.